gio/src/subclass/action.rs

```diff
@@ -0,0 +1,417 @@
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface =
+                type_data.as_ref().parent_interface::<Action>() as *const ffi::GActionInterface;
+
+            let func = (*parent_iface)
+                .activate
+                .expect("no parent \"activate\" implementation");
+            func(
+                self.obj().unsafe_cast_ref::<Action>().to_glib_none().0,
+                parameter.to_glib_none().0,
+            );
+        }
+    }
+
+    fn delegate_get_property(
+        &self,
+        prop_id: usize,
+        _pspec: &glib::ParamSpec,
+    ) -> Option<glib::Value> {
+        let type_: glib::Type = self.obj().type_();
+        let property = ActionProperty::from_type(type_, prop_id)?;
+        Some(match property {
+            ActionProperty::Name => self.name().to_value(),
+            ActionProperty::ParameterType => self.parameter_type().to_value(),
+            ActionProperty::StateType => self.state_type().to_value(),
+            ActionProperty::Enabled => self.is_enabled().to_value(),
+            ActionProperty::State => self.state().to_value(),
+        })
+    }
+
+    fn delegate_set_property(
+        &self,
+        _prop_id: usize,
+        _value: &glib::Value,
+        _pspec: &glib::ParamSpec,
+    ) -> bool {
+        false
+    }
+}
+
+impl<T: ActionImpl> ActionImplExt for T {}
+
+static FIRST_PROPERTY_ID_QUARK: LazyLock<glib::Quark> =
+    LazyLock::new(|| glib::Quark::from_str("gtk-rs-subclass-action-first-prop"));
+
+fn set_first_prop(type_: glib::Type, first_prop: usize) {
+    let data = Box::new(first_prop);
+    unsafe {
+        glib::gobject_ffi::g_type_set_qdata(
+            type_.into_glib(),
+            FIRST_PROPERTY_ID_QUARK.into_glib(),
+            Box::into_raw(data) as *mut _,
+        );
+    }
+}
+
+fn get_first_prop(type_: glib::Type) -> Option<usize> {
+    unsafe {
+        let ptr = glib::gobject_ffi::g_type_get_qdata(
+            type_.into_glib(),
+            FIRST_PROPERTY_ID_QUARK.into_glib(),
+        ) as *mut usize;
+        if ptr.is_null() {
+            None
+        } else {
+            Some(*ptr)
+        }
+    }
+}
+
+#[repr(C)]
+enum ActionProperty {
+    Name = 0,
+    ParameterType,
+    StateType,
+    Enabled,
+    State,
+}
+
+impl ActionProperty {
+    fn from_id(first_prop: usize, id: usize) -> Option<Self> {
+        match id.checked_sub(first_prop)? {
+            0 => Some(Self::Name),
+            1 => Some(Self::ParameterType),
+            2 => Some(Self::StateType),
+            3 => Some(Self::Enabled),
+            4 => Some(Self::State),
+            _ => None,
+        }
+    }
+
+    fn from_type(mut type_: glib::Type, id: usize) -> Option<Self> {
+        loop {
+            if let Some(first_prop) = get_first_prop(type_) {
+                break ActionProperty::from_id(first_prop, id);
+            }
+            type_ = type_.parent()?;
```

Comment:
Is this actually correct to look at all the parents, can you add a comment why?
Shouldn't this just look at the just the type of `T`?

Maybe a test with multiple levels of `GAction` inheritance would be useful to make sure this works and continues working